File: Semester 6/PO/Listy/Lista 1/zad-3-4.py
class LinkedList:
    """Class representing a singly linked list."""

    class Node:
        def __init__(self, value):
            self.value = value
            self.next = None

    def __init__(self, values=None):
        """Initializes the linked list with optional values."""
        self.head = None
        if values:
            for value in reversed(values):
                self.prepend(value)

    def prepend(self, value):
        """Adds a value to the start of the list."""
        node = self.Node(value)
        node.next = self.head
        self.head = node

    def map_procedural(self, func):
        """Creates a new list with function applied to each element (procedural)."""
        result = LinkedList()
        current = self.head
        while current:
            result.prepend(func(current.value))  # Using prepend to maintain order
            current = current.next
        # Convert to list, reverse, and create new LinkedList
        values = list([node.value for node in result])
        return LinkedList(values[::-1])

    def map_functional(self, func):
        """Returns a new linked list with function applied (functional)."""
        return LinkedList([func(node.value) for node in self])

    def filter_procedural(self, predicate):
        """Creates a new list with elements that match predicate (procedural)."""
        result = LinkedList()
        current = self.head
        while current:
            if predicate(current.value):
                result.prepend(current.value)  # Using prepend to maintain order
            current = current.next
        # Convert to list, reverse, and create new LinkedList
        values = list([node.value for node in result])
        return LinkedList(values[::-1])

    def filter_functional(self, predicate):
        """Returns a new linked list with only elements matching predicate (functional)."""
        return LinkedList([node.value for node in self if predicate(node.value)])

    def __iter__(self):
        """Iterator to traverse the linked list."""
        current = self.head
        while current:
            yield current
            current = current.next

    def __repr__(self):
        """String representation of the linked list."""
        return ' -> '.join(str(node.value) for node in self) or "Empty List"

if __name__ == "__main__":
    linked_list = LinkedList([1, 2, 3, 4, 5])
    print("Original List:", linked_list)

    squared_list = linked_list.map_functional(lambda x: x ** 2)
    print("Squared List (functional):", squared_list)

    added_list = linked_list.map_procedural(lambda x: x + 2)
    print("Added 2 List (procedural):", added_list)

    filtered_even = linked_list.filter_functional(lambda x: x % 2 == 0)
    print("Even numbers (functional):", filtered_even)

    filtered_odd = linked_list.filter_procedural(lambda x: x % 2 != 0)
    print("Odd numbers (procedural):", filtered_odd)

File: Semester 6/PO/Listy/Lista 1/zad-3-4.c
#include <stdio.h>
#include <stdlib.h>

// Definition of a singly linked list node
typedef struct Node {
    int value;
    struct Node *next;
} Node;

// List creation function
Node* create_list() {
    return NULL;
}

// Add value to list function
Node* add_value(Node* head, int value) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    if(!new_node) return head;  // Handle allocation failure

    new_node->value = value;
    new_node->next = NULL;

    if(!head) {
        return new_node;  // First node
    }

    // Find the last node
    Node* current = head;
    while(current->next) {
        current = current->next;
    }
    current->next = new_node;
    return head;
}

// Procedural version of map function
Node* map_procedural(Node *head, int (*func)(int)) {
    Node *new_head = NULL;
    Node *current = head;

    while(current) {
        // Create new node with transformed value
        Node *new_node = (Node*)malloc(sizeof(Node));
        if(!new_node) return new_head;  // Handle allocation failure

        new_node->value = func(current->value);
        new_node->next = NULL;

        if(!new_head) {
            new_head = new_node;  // First node
        }

        else {
            // Find the last node
            Node *tmp = new_head;
            while(tmp->next) {
                tmp = tmp->next;
            }
            tmp->next = new_node;
        }
        current = current->next;
    }
    return new_head;
}

// Functional version of map function
Node* map(Node *head, int (*func)(int)) {
    if(!head) return NULL;
    Node *new_head = (Node*)malloc(sizeof(Node));
    new_head->value = func(head->value);
    new_head->next = map(head->next, func);
    return new_head;
}

// Example functions to apply
int square(int x) {
    return x * x;
}

int add_two(int x) {
    return x + 2;
}

// Procedural version of filter function
Node* filter_procedural(Node *head, int (*predicate)(int)) {
    Node *new_head = NULL;
    Node *current = head;

    while(current) {
        if(predicate(current->value)) {
            Node *new_node = (Node*)malloc(sizeof(Node));
            if(!new_node) return new_head;  // Handle allocation failure

            new_node->value = current->value;
            new_node->next = NULL;

            if(!new_head) {
                new_head = new_node;  // First node
            }

            else {
                // Find the last node
                Node *tmp = new_head;
                while(tmp->next) {
                    tmp = tmp->next;
                }
                tmp->next = new_node;
            }
        }
        current = current->next;
    }
    return new_head;
}

// Functional version of filter function
Node* filter(Node *head, int (*predicate)(int)) {
    if(!head) return NULL;
    if(predicate(head->value)) {
        Node *new_node = (Node*)malloc(sizeof(Node));
        new_node->value = head->value;
        new_node->next = filter(head->next, predicate);
        return new_node;
    }
    return filter(head->next, predicate);
}

// Example predicate functions
int is_even(int x) {
    return x % 2 == 0;
}

int is_odd(int x) {
    return x % 2 != 0;
}

// Function to print linked list
void print_list(Node *head) {
    while(head) {
        printf("%d ", head->value);
        head = head->next;
    }
    printf("\n");
}

// Function to free linked list memory
void free_list(Node *head) {
    while(head) {
        Node *tmp = head;
        head = head->next;
        free(tmp);
    }
}

int main() {
    Node* list = create_list();
    list = add_value(list, 1);
    list = add_value(list, 2);
    list = add_value(list, 3);
    list = add_value(list, 4);
    list = add_value(list, 5);

    printf("Original list: ");
    print_list(list);

    Node* squared_list = map(list, square);
    printf("Squared list (functional): ");
    print_list(squared_list);

    Node* added_list = map_procedural(list, add_two);
    printf("Added 2 list (procedural): ");
    print_list(added_list);

    Node* filtered_list = filter(list, is_even);
    printf("Even numbers list (functional): ");
    print_list(filtered_list);

    Node* filtered_list_procedural = filter_procedural(list, is_odd);
    printf("Odd numbers list (procedural): ");
    print_list(filtered_list_procedural);

    // Free all remaining lists
    free_list(list);
    free_list(squared_list);
    free_list(added_list);
    free_list(filtered_list);
    free_list(filtered_list_procedural);
    return 0;
}